### ISSUE TITLE
warm_start_from doesn't initialize variables from the checkpoint

### ISSUE THREAD

**Issue by happylittlebunny** [tensorflow/tensor2tensor#1265]
### Description

I am trying to do transfer learning on top of the tensor2tensor translation problem.
I use warm_start_from param. The code shows it triggers initialize_from_ckpt and shall be supposed to init variables based on my checkpoint(which is my base model) passed in the warm_start_from param . However when I checked the new model's initial checkpoint model.ckpt-0, I found all the variables' value are 0. So the initialization still assigns the variable 0. Or the initialization is overwritten somewhere? 

The interesting thing is since I freezes decoder by not passing decoder's variable into optimize_loss function. The decoder's variable is retained. So except decoder tensor's variables are initialized from base model. All other tensors' initial value are 0.

I am not sure if it's a bug or a usage problem.
How can I get other variables to be initialized with the value from a checkpoint as well?

### Environment information

```
OS: win10

$ pip freeze | grep tensor
# your output here

$ python -V
# your output here
```

### For bugs: reproduction and error logs

```
# Steps to reproduce:
...
```

```
# Error logs:
...
```

**Comment by ybbaigo** [comment]
Could your provide your training command here?

**Comment by ybbaigo** [comment]
So you are using `transformer` model, according to your hparams `transformer_base_single_gpu`, some dropout layers are enabled, for example

    hparams.attention_dropout = 0.1
    hparams.relu_dropout = 0.1

So I think the inconsistency is in line with expectations.  @happylittlebunny

**Comment by happylittlebunny** [comment]
thanks @ybbaigo for the explanation. not sure if it's a stupid question. Can I set random seed in the t2t code to produce stable result ?

**Comment by ybbaigo** [comment]
You can disable such layers through hparam_set, but this will affect the generalization ability of the model. @happylittlebunny

**Comment by afrozenator** [comment]
There is a way to set the random seed and we do it right in the beginning
in t2t_train if you follow the code you'll get it.

Can you try to add it again after / just before where we do warm_start and
see if it fixes your issue, rather than disabling dropout layers.

@happybunny

**Comment by happylittlebunny** [comment]
thanks @afrozenator  for pointing it out. I added param --random_seed=888 in t2t_train, and I checked the random seed is set in the flag.txt. But I still get different loss and model each time I ran with the same params and the same random seed...
Is it because the tf.set_random_seed() can't control the dropout result ?

**Comment by afrozenator** [comment]
I meant change the code to set the random seed again once we restore from
the checkpoint -- that said the exact same loss may not be possible to
achieve, however one should try to get it close

**Comment by happylittlebunny** [comment]
I have added 'tf.set_random_seed(888)' in t2t_model.py right after the line:

if self._hparams.warm_start_from:
     self.initialize_from_ckpt(self._hparams.warm_start_from)
    tf.set_random_seed(888)


From the loss I compared from different runs(same params) and the decoder results and bleu score results. I don't think they produce the same or similar model.. I

I searched up online.  It seems there are some tf operations are GPU non-deterministic , ie reduce_sum. I am not sure if it's fixed or not. Do you know ? If it's not fixed or can't be fixed. I guess that's why it can't produce the same or similar model?

**Comment by afrozenator** [comment]
That makes sense, but it might apply more in a distributed setting rather
than single CPU setting. (That is another idea, try without GPUs)

But I think, as long as the results are similar -- it should be fine, after
all ensembling works exactly due to this reason only!
